File: frontend/packages/types/src/api/onboarding-status.ts
import type { AuthMethodKind, CountryCode, CustomDocumentUploadSettings, DocumentRequestKind } from '..';
import type {
  CollectedDataOption,
  CollectedInvestorProfileDataOption,
  CollectedKybDataOption,
  CollectedKycDataOption,
} from '../data/collected-data-option';
import type { SupportedIdDocTypes } from '../data/id-doc-type';

/** Note: these will be a little trickier to migrate to the open API type. The open API type has a new, more modern serialization */
export enum OnboardingRequirementKind {
  registerAuthMethod = 'register_auth_method',
  registerPasskey = 'liveness', // TODO: eventually we need to synchronously rename `liveness` or create a duplicate
  document = 'collect_document',
  collectKycData = 'collect_data',
  collectKybData = 'collect_business_data',
  createBusinessOnboarding = 'create_business_onboarding',
  investorProfile = 'collect_investor_profile',
  authorize = 'authorize',
  process = 'process',
}

/** Note: these will be a little trickier to migrate to the open API type. The open API type has a new, more modern serialization */
export type RegisterAuthMethodRequirement = {
  kind: OnboardingRequirementKind.registerAuthMethod;
  isMet: boolean;
  authMethodKind: AuthMethodKind;
};

/** Note: these will be a little trickier to migrate to the open API type. The open API type has a new, more modern serialization */
export type CreateBusinessOnboardingRequirement = {
  kind: OnboardingRequirementKind.createBusinessOnboarding;
  requiresBusinessSelection: boolean;
};

/** Note: these will be a little trickier to migrate to the open API type. The open API type has a new, more modern serialization */
export type CollectKybDataRequirement = {
  kind: OnboardingRequirementKind.collectKybData;
  isMet: boolean;
  missingAttributes: CollectedKybDataOption[];
  populatedAttributes: CollectedKybDataOption[];
  recollectAttributes: CollectedKybDataOption[];
};

/** Note: these will be a little trickier to migrate to the open API type. The open API type has a new, more modern serialization */
export type CollectKycDataRequirement = {
  kind: OnboardingRequirementKind.collectKycData;
  isMet: boolean;
  missingAttributes: CollectedKycDataOption[];
  populatedAttributes: CollectedKycDataOption[];
  recollectAttributes: CollectedKybDataOption[];
  optionalAttributes: CollectedKycDataOption[];
};

/** Note: these will be a little trickier to migrate to the open API type. The open API type has a new, more modern serialization */
export type CollectInvestorProfileRequirement = {
  kind: OnboardingRequirementKind.investorProfile;
  isMet: boolean;
  missingAttributes: CollectedInvestorProfileDataOption[];
  missingDocument?: boolean;
  populatedAttributes: CollectedInvestorProfileDataOption[];
};

/** Note: these will be a little trickier to migrate to the open API type. The open API type has a new, more modern serialization */
export type DocumentRequirement<TConfig = DocumentRequirementConfig> = {
  kind: OnboardingRequirementKind.document;
  isMet: boolean;
  documentRequestId: string;
  uploadSettings: DocumentUploadSettings;
  config: TConfig;
};

export enum DocumentUploadSettings {
  /**
   * Useful for documents that are usually uploaded as a captured image, like an SSN card or
   * driver's license. When on desktop, we prefer to hand off these document requirements to
   * mobile for capture. And on mobile, we will show the capture interface with an option to
   * upload.
   */
  preferCapture = 'prefer_capture',
  /**
   * Useful for documents that are usually uploaded as a file image, like a lease or utility
   * bill. When on desktop, we first give the option to upload on desktop but fall back to
   * handoff to mobile. And on mobile, we will show the option to upload or capture.
   */
  preferUpload = 'prefer_upload',
  /**
   * Not configurable by tenants yet, only used by Coba for their identity documents.
   * On mobile, only allows capturing without the option to upload.
   * On desktop though, we will allow uploading if the user didn't hand off
   */
  captureOnlyOnMobile = 'capture_only_on_mobile',
}

export type IdDocRequirementConfig = {
  kind: DocumentRequestKind.Identity;
  shouldCollectSelfie: boolean;
  shouldCollectConsent: boolean;
  supportedCountryAndDocTypes: Partial<Record<CountryCode, SupportedIdDocTypes[]>>;
};

export type ProofOfAddressRequirementConfig = {
  kind: DocumentRequestKind.ProofOfAddress;
};

export type ProofOfSsnRequirementConfig = {
  kind: DocumentRequestKind.ProofOfSsn;
};

export type CustomDocumentRequirementConfig = {
  kind: DocumentRequestKind.Custom;
  name: string;
  description?: string;
  uploadSettings: CustomDocumentUploadSettings;
};

export type DocumentRequirementConfig =
  | IdDocRequirementConfig
  | ProofOfAddressRequirementConfig
  | ProofOfSsnRequirementConfig
  | CustomDocumentRequirementConfig;

/** Note: these will be a little trickier to migrate to the open API type. The open API type has a new, more modern serialization */
export type RegisterPasskeyRequirement = {
  kind: OnboardingRequirementKind.registerPasskey;
  isMet: boolean;
};

/** Note: these will be a little trickier to migrate to the open API type. The open API type has a new, more modern serialization */
export type AuthorizeRequirement = {
  kind: OnboardingRequirementKind.authorize;
  isMet: boolean;
  fieldsToAuthorize: AuthorizeFields;
};

/** Note: these will be a little trickier to migrate to the open API type. The open API type has a new, more modern serialization */
export type ProcessRequirement = {
  kind: OnboardingRequirementKind.process;
  isMet: boolean;
};

/** Note: these will be a little trickier to migrate to the open API type. The open API type has a new, more modern serialization */
export type OnboardingRequirement =
  | RegisterAuthMethodRequirement
  | CollectKybDataRequirement
  | CollectKycDataRequirement
  | CollectInvestorProfileRequirement
  | DocumentRequirement
  | RegisterPasskeyRequirement
  | AuthorizeRequirement
  | ProcessRequirement;

export type OnboardingStatusRequest = {
  authToken: string;
};

export type AuthorizeFields = {
  collectedData: CollectedDataOption[];
  documentTypes: SupportedIdDocTypes[];
};

export type OnboardingStatusResponse = {
  allRequirements: OnboardingRequirement[];
  canUpdateUserData: boolean;
};

export type RequirementForKind<K> = K extends OnboardingRequirementKind.registerPasskey
  ? RegisterPasskeyRequirement
  : K extends OnboardingRequirementKind.document
    ? DocumentRequirement
    : K extends OnboardingRequirementKind.collectKycData
      ? CollectKycDataRequirement
      : K extends OnboardingRequirementKind.collectKybData
        ? CollectKybDataRequirement
        : K extends OnboardingRequirementKind.investorProfile
          ? CollectInvestorProfileRequirement
          : K extends OnboardingRequirementKind.authorize
            ? AuthorizeRequirement
            : K extends OnboardingRequirementKind.process
              ? ProcessRequirement
              : never;

export const getRequirement = <K extends OnboardingRequirementKind>(reqs: OnboardingRequirement[], kind: K) =>
  getRequirements(reqs, kind)[0];

export const getRequirements = <K extends OnboardingRequirementKind>(reqs: OnboardingRequirement[], kind: K) => {
  const found = reqs.filter(req => req.kind === kind);
  return found as RequirementForKind<K>[];
};

File: mobile/packages/footprint_flutter/lib/src/onboarding-components/models/onboarding_status.dart
import 'package:footprint_flutter/src/onboarding-components/models/collected_data_option.dart';

enum OnboardingRequirementKind {
  registerAuthMethod("register_auth_method"),
  registerPasskey("liveness"),
  document("collect_document"),
  collectKycData("collect_data"),
  collectKybData("collect_business_data"),
  investorProfile("collect_investor_profile"),
  authorize("authorize"),
  process("process");

  final String value;

  const OnboardingRequirementKind(this.value);

  static OnboardingRequirementKind? fromValue(String value) {
    switch (value) {
      case "register_auth_method":
        return OnboardingRequirementKind.registerAuthMethod;
      case "liveness":
        return OnboardingRequirementKind.registerPasskey;
      case "collect_document":
        return OnboardingRequirementKind.document;
      case "collect_data":
        return OnboardingRequirementKind.collectKycData;
      case "collect_business_data":
        return OnboardingRequirementKind.collectKybData;
      case "collect_investor_profile":
        return OnboardingRequirementKind.investorProfile;
      case "authorize":
        return OnboardingRequirementKind.authorize;
      case "process":
        return OnboardingRequirementKind.process;
      default:
        return null;
    }
  }

  @override
  String toString() {
    return value;
  }
}

class Requirement {
  final OnboardingRequirementKind kind;
  final List<CollectedDataOption>? missingAttributes;
  final List<CollectedDataOption>? populatedAttributes;
  final List<CollectedDataOption>? optionalAttributes;

  Requirement({
    required this.kind,
    this.missingAttributes,
    this.populatedAttributes,
    this.optionalAttributes,
  });

  factory Requirement.fromJson(Map<String, dynamic> json) {
    List<CollectedDataOption>? missingAttributes = [];
    List<CollectedDataOption>? populatedAttributes = [];
    List<CollectedDataOption>? optionalAttributes = [];

    if (json['missing_attributes'] != null) {
      for (var element in (json['missing_attributes'] as List<dynamic>)) {
        if (CollectedDataOption.fromValue(element) != null) {
          missingAttributes.add(CollectedDataOption.fromValue(element)!);
        }
      }
    }

    if (json['populated_attributes'] != null) {
      for (var element in (json['populated_attributes'] as List<dynamic>)) {
        if (CollectedDataOption.fromValue(element) != null) {
          populatedAttributes.add(CollectedDataOption.fromValue(element)!);
        }
      }
    }

    if (json['optional_attributes'] != null) {
      for (var element in (json['optional_attributes'] as List<dynamic>)) {
        if (CollectedDataOption.fromValue(element) != null) {
          optionalAttributes.add(CollectedDataOption.fromValue(element)!);
        }
      }
    }

    return Requirement(
      kind: OnboardingRequirementKind.fromValue(json['kind'])!,
      missingAttributes:
          json['missing_attributes'] != null ? missingAttributes : null,
      populatedAttributes:
          json['populated_attributes'] != null ? populatedAttributes : null,
      optionalAttributes:
          json['optional_attributes'] != null ? optionalAttributes : null,
    );
  }
}

// Notice that this an oversimplified version of the original model in "frontend/packages/types/src/api/onboarding-status.ts"
// This is because Dart does not have union types
// So, it would be quite complex to implement the original model in Dart
// This model should serve the purpose for now
class OnboardingRequirement {
  final bool isMet;
  final Requirement requirement;

  OnboardingRequirement({
    required this.isMet,
    required this.requirement,
  });

  factory OnboardingRequirement.fromJson(Map<String, dynamic> json) {
    return OnboardingRequirement(
      isMet: json['is_met'],
      requirement: Requirement.fromJson(json),
    );
  }
}

class OnboardingStatusResponse {
  final List<OnboardingRequirement> allRequirements;
  final bool canUpdateUserData;

  OnboardingStatusResponse({
    required this.allRequirements,
    required this.canUpdateUserData,
  });

  factory OnboardingStatusResponse.fromJson(Map<String, dynamic> json) {
    List<OnboardingRequirement> allRequirements = [];
    bool canUpdateUserData = false;

    if (json['all_requirements'] != null) {
      for (var element in (json['all_requirements'] as List<dynamic>)) {
        allRequirements.add(OnboardingRequirement.fromJson(element));
      }
    }
    if (json['can_update_user_data'] != null) {
      canUpdateUserData = json['can_update_user_data'];
    }

    return OnboardingStatusResponse(
      allRequirements: allRequirements,
      canUpdateUserData: canUpdateUserData,
    );
  }
}
